Improve API

simonepri opened the pull request:
## Changes
- Add CLI
- Improve performance
- Improve documentation
- If the advanced options is provided it returns an array of `{pid, ppid}` instead of an array of `pids` 
- Fix handling of empty lines in stdout

Preview here: https://github.com/simonepri/pidtree/tree/refactor

@zixia @soyuka what do you think about the [CLI](https://github.com/simonepri/pidtree/tree/refactor#cli)?

soyuka commented on bin/pidtree.js:
```diff
@@ -3,19 +3,133 @@
+      return;
+    }
+
+    var fathers = {}; // HasMap of fathers
```
parents? Father is weird but I get it :smile:

simonepri replied:
Thanks 🙈in italian is called `vettore dei padri` where `padri` = `fathers`.
But in english is called `vector of parents` actually.  😅

soyuka commented on bin/pidtree.js:
```diff
@@ -3,19 +3,133 @@
+      }
+      delete tree[cur.pid];
+    }
+    return root;
```
Lots of loops no? I don't have much time to study this but looks like there's something simpler to be done. For example by already printing above (not sure)?

simonepri replied:
This is just an iterative version of the DFS algorithm that I use to create the tree. Quite standard.
Obviously I can print the tree directly without actually creating the tree but I was tired yesterday.

soyuka replied:
> but I was tired yesterday.

I think it reflected in your code :p.

soyuka commented on lib/pidtree.js:
```diff
@@ -0,0 +1,82 @@
+        pids.push({ppid: curpid, pid: tree[curpid][j]});
+      }
+      delete tree[curpid];
+    }
```
Can't those two steps be made in one pass?

simonepri replied:
The first step builds the [adjacency list](https://en.wikipedia.org/wiki/Adjacency_list)
Then after that I start visiting the tree from the root node using the a [BFS](https://en.wikipedia.org/wiki/Breadth-first_search).

BUT I can build the vector of parents instead of an adjacency list.
But then I should do a leaf to parent visit. Trickier to write but should be faster.

simonepri replied:
This is the implementation using the vector of parents.
Maybe is better to just keep the old one, should be easier to maintain.
This should be a bit more memory efficient since in general pid trees are more wider than deeper but practically there isn't such a big difference.
```js
// Build the HashMap of parents (pid -> ppid)
    var info = {};
    var pids = [];
    while (list.length > 0) {
      var e = list.pop();
      info[e[1]] = {parent: e[0]};
      pids.push(e[1]);
    }

    // For each pid in the system:
    //  - Traverse the path from leaf to root until:
    //    - the info of whether or not the current pid is grandchildren of the
    //      PID the user is interested in is already known.
    //    - OR
    //    - the current pid is a root.
    //  - Mark each node in the path from leaf to root with the info of
    //    whether or not they are grandchildren of the PID the user is
    //    interested.
    while (pids.length > 0) {
      var last = pids.pop();
      var path = [];
      var status = false;
      // eslint-disable-next-line no-constant-condition
      while (true) {
        if (!info[last]) {
          info[last] = {};
        }
        var data = info[last];
        // Insert the pid in the path.
        path.push(last);
        if (last === PID || data.ancestor === true) {
          // This pid is the one the user is interested in or is one of its
          // grandchildren
          status = true;
          break;
        }
        if (data.ancestor === false || data.parent === undefined) {
          // This pid in not a grandchildren of the PID the user is interested in
          // or this is a root.
          break;
        }
        last = data.parent;
      }
      // Mark all the PIDS in the path traversed with the info of whether or not
      // they are grandchildren of the PID the user is interested.
      if (info[path[path.length - 1]] === undefined) {
        info[path[path.length - 1]] = {};
      }
      while (path.length > 0) {
        info[path.pop()].ancestor = status;
      }
    }

    // Do a linear scan to get the list of pids that are children or
    // grandchildren of the PID the user is interested.
    // eslint-disable-next-line guard-for-in
    for (var pid in info) {
      if (info[pid].ancestor) {
        pids.push(
          options.advanced
            ? {ppid: info[pid].parent, pid: Number(pid)}
            : Number(pid)
        );
      }
      delete info[pid];
    }

    if (!options.root) {
      pids.shift(); // Remove root
    }
```

soyuka commented:
tree is swag \o/

soyuka commented:
looks really good, sorry I couldn't review more had a really busy day!

simonepri commented:
@soyuka many thanks for the review.
I've removed the breaking change and added the 'advanced' option.

I saw this:
https://github.com/soyuka/pidusage-tree/issues/5

Have you found a faster way to get the list?

soyuka commented:
https://github.com/soyuka/pidusage-tree/blob/master/index.js#L1 done actually \o/

simonepri commented:
Oh `ptree` is `pidtree`? Didn't get that.
